### PR TITLE
Improve MA0042 async overload matching coverage

### PR DESCRIPTION
**Pull request by meziantou**
## Why
`MA0042` overload detection was missing several realistic call-shape scenarios (generic inference, extension-vs-instance matching, ref-kind nuances, and numeric conversion compatibility), which caused false negatives in async context diagnostics.

## What changed
- Expanded overload matching in `OverloadFinder` to better support:
  - method-generic inference/constraint validation during comparison
  - `ref`/`out` strictness with `in`/by-value compatibility
  - extension/instance comparable parameter normalization
  - safer implicit numeric conversion compatibility (including `System.Half -> float/double`)
- Improved async-equivalent candidate discovery in `DoNotUseBlockingCallInAsyncContextAnalyzer` by broadening lookup around receiver/type context.
- Added a large async-context test matrix covering:
  - generic compatibility (positive and negative)
  - numeric widening/narrowing and floating-point precision-sensitive cases
  - instance vs extension method matching
  - argument forms (`default`, `null`, named args) and modifier edge cases
- Updated `DocumentationGenerator` to use `FindRequiredGitRepositoryRoot()` and bumped `Meziantou.Framework.FullPath` to `1.1.18`.

## Validation
- `dotnet test --filter "FullyQualifiedName~DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.Argument_ImplicitNumeric"` (passed)
- `dotnet test --filter "FullyQualifiedName~DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests"` (passed)
- `dotnet run --project src/DocumentationGenerator` currently fails in this worktree with `Cannot find the current git folder` in this environment; code was updated toward required FullPath usage.

### CODE REVIEW

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs [hunk 1]
```diff
@@ -812,23 +808,964 @@ public async Task a()
...
+                {
+                    public void A(List<int> value) => throw null;
+                    public Task AAsync<T>(IEnumerable<T> value, CancellationToken token = default) => throw null;
+                }
+
+                class Demo
+                {
+                    public async Task M()
+                    {
+                        await new Test().AAsync(new List<int>());
```
Comment (meziantou) [inline tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs]: Should it report? List is assignable to IEnumerable, but maybe the sync method is more optimized for List.

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs [hunk 2]
```diff
@@ -812,23 +808,964 @@ public async Task a()
...
+                    public void A(List<List<int>> value) => throw null;
+                    public Task AAsync<T>(List<List<T>> value, CancellationToken token = default) => throw null;
+                    public Task AAsync(List<List<string>> value, CancellationToken token = default) => throw null;
+                }
+
+                class Demo
+                {
+                    public async Task M()
+                    {
+                        new Test().A(new List<List<int>>());
```
Comment (meziantou) [inline tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:933]: Should it report?

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs [hunk 3]
```diff
@@ -812,23 +808,964 @@ public async Task a()
...
+                    public void A<T>(List<T> value) => throw null;
+                    public Task AAsync<T>(List<T> value, CancellationToken token = default)
+                        where T : class => throw null;
+                }
+
+                class Demo
+                {
+                    public async Task M()
+                    {
+                        new Test().A<int>(new List<int>());
```
Comment (meziantou) [inline tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:1127]: Should it report if `A<string>`?

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs [hunk 4]
```diff
@@ -812,23 +808,964 @@ public async Task a()
...
+        await CreateProjectBuilder()
+              .WithSourceCode("""
+                using System.Collections.Generic;
+                using System.Threading;
+                using System.Threading.Tasks;
+
+                class Test
+                {
+                    public void A(List<int> value) => throw null;
+                    public Task AAsync<T>(IReadOnlyCollection<T> value, CancellationToken token = default) => throw null;
```
Comment (meziantou) [inline tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:1599]: Add a test with IEnumerable
```
                    public void A(List<int> value) => throw null;
                    public Task AAsync<T>(IEnumerable<T> value, CancellationToken token = default) => throw null;
```

I don't think it should report in this case.

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs [hunk 5]
```diff
@@ -812,23 +808,1008 @@ public async Task a()
...
+    public async Task Argument_InModifierDifference_ShouldReport()
+    {
+        await CreateProjectBuilder()
+              .WithSourceCode("""
+                using System.Threading;
+                using System.Threading.Tasks;
+
+                class Test
+                {
+                    public void A(in int value) => throw null;
```
Comment (Griboedoff) [inline tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:1144]: may be there should be separate analyzer without fix for `in`, `ref` and `out`, because modifiers are forbidden for async methods but you can rewrite it to async version

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs [hunk 6]
```diff
@@ -812,23 +808,1008 @@ public async Task a()
...
+                    {
+                        [|new Test().A((Half)1)|];
+                    }
+                }
+                """)
+              .ValidateAsync();
+    }
+
+    [Fact]
+    public async Task GenericArgument_NestedCompatibleGenericDefinitions_ShouldReport()
```
Comment (Griboedoff) [inline tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs]: `NestedCompatibleGenericDefinitions`

not nested, just compatible

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs [hunk 7]
```diff
@@ -812,23 +808,1008 @@ public async Task a()
...
+    {
+        await CreateProjectBuilder()
+              .WithSourceCode("""
+                using System.Collections.Generic;
+                using System.Threading;
+                using System.Threading.Tasks;
+
+                class Test
+                {
+                    public void A(Dictionary<int, string> value) => throw null;
```
Comment (Griboedoff) [inline tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:1624]: arguments are incompatible without arity

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs [hunk 8]
```diff
@@ -812,23 +808,1008 @@ public async Task a()
...
+
+                class Value
+                {
+                    public static implicit operator Value(string value) => new Value();
+                }
+
+                class Test
+                {
+                    public void A(Value value) => throw null;
+                    public Task AAsync(Value value, CancellationToken token = default) => throw null;
```
Comment (Griboedoff) [inline tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:1751]: parameters are compatible without implicit cast, may be sync method should be `A(string value)`

File: tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs (hunk 5 above)
Comment (meziantou) [reply tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:1144]: The code should only accept the `in` modifier. `out` and `ref` would be incompatible. While it's an invalid modifier when you define an `async` method, an async method can use a method with a `in` parameter.

As the use-case is to replace method invocation, I think this is ok.
```c#
Task Foo(in int a) => throw null!;

async Task Bar(int a)
{
    await Foo(a); // Valid
}
```

Comment (Griboedoff) [reply tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:1144]: this sounds reasonable


But I'm talking about other case
```
bool TryA(int b, out Foo f) => throw null!;
Task(bool, Foo)> TryAAsync(int b) => throw null!;
```

I understand that it's a long shot for analyzer to guess this type of overload, but it can be usefull :)

Comment (meziantou) [reply tests/Meziantou.Analyzer.Test/Rules/DoNotUseBlockingCallInAsyncContextAnalyzer_AsyncContextTests.cs:1144]: It would have more impact that just finding the _overload_. The code fix must be aware of it, not just using another method name, but assigning variables (it can be tricky depending on the context).
